src/monattack.cpp

```diff
@@ -4305,14 +4316,452 @@ void mattack::suicide(monster *z, int index)
+
+    // attack types
+    enum : int {
+        att_grenade,
+        att_flashbang,
+        att_gasbomb,
+        att_c4,
+        att_mininuke,
+        att_enum_size
+    };
+
+    // max distance that hacks can be release towards the target.
+    constexpr auto max_distance = 30;
+
+    // attack movement costs
+    constexpr int att_cost = 60;
+
+    if( !z->can_act() ) {
+        return;
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    // Look for a valid target...
+
+    // Only can actively target the player right now. Once we have the ability to grab targets that we aren't
+    // actively attacking change this to use that instead.
+    Creature *const target = static_cast<Creature *>(&g->u);
+    if (z->attitude_to( *target ) == Creature::A_FRIENDLY) {
+        return;
+    }
+
+    // too far or can't target them
+    if (!within_target_range(z, target, max_distance)) {
+        return;
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    // ok, we have a valid target; populate valid attack options...
+    std::array<int, att_enum_size> valid_attacks;
+    size_t valid_attack_count = 0;
+
+    // need an open space for these attacks
+    auto const empty_neighbors = find_empty_neighbors(*z);
+    size_t const empty_neighbor_count = empty_neighbors.second;
+    if (!empty_neighbor_count) {
+        return;
+    }
+
+    if( z->ammo["bot_grenade_hack"] > 0 ) {
+        valid_attacks[valid_attack_count++] = att_grenade;
+    }
+    if( z->ammo["bot_flashbang_hack"] > 0 ) {
+        valid_attacks[valid_attack_count++] = att_flashbang;
+    }
+    if( z->ammo["bot_gasbomb_hack"] > 0 ) {
+        valid_attacks[valid_attack_count++] = att_gasbomb;
+    }
+    // Check this one later, suffers no chance penalty if it and mininukes are the only ones left
+    if( z->ammo["bot_c4_hack"] > 0  && (one_in(2) || valid_attack_count == 0)) {
+        valid_attacks[valid_attack_count++] = att_c4;
+    }
+    // Check this one last, suffers no chance penalty if it's the only one left
+    if( z->ammo["bot_mininuke_hack"] > 0  && (one_in(50) || valid_attack_count == 0)) {
+        valid_attacks[valid_attack_count++] = att_mininuke;
+    }
+
+    if (valid_attack_count == 0) {
+        // We've run out of ammo, time to YOLO!
+        z->anger = 100;
+        // Toggle off this special.
+        z->set_special(index, -1);
+        return;
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    // choose and do a valid attack
+    const int attack_index = get_random_index(valid_attack_count);
+    const point where = empty_neighbors.first[get_random_index(empty_neighbor_count)];
+    switch (valid_attacks[attack_index]) {
+    default :
+        DebugLog(D_WARNING, D_GAME) << "Bad enum value in grenadier.";
+        break;
+    case att_grenade : {
```

Comment:
The code duplication is strong in this one (most of the PR), but in this switch it's especially obvious: each case is the same, only different monster type id, message and ammo item type id.